spring-boot-modules/spring-boot-crud/src/test/java/com/baeldung/entity_dto_differences/controller/UserControllerIntegrationTest.java

```diff
@@ -0,0 +1,92 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+
+import java.util.Arrays;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+public class UserControllerIntegrationTest {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Autowired
+    private UserRepository userRepository;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeEach
+    public void beforeEach() {
+
+        userRepository.deleteAll();
+    }
+
+    @Test
+    public void whenGetUsers_thenCorrect() throws Exception {
+
+        //given
+        Book book1 = new Book("Book1", "Author1");
+        Book book2 = new Book("Book2", "Author2");
+        User savedUser = userRepository.save(new User("John", "Doe", "123 Main St", Arrays.asList(book1, book2)));
+
+        //when
+        MvcResult result = mockMvc.perform(MockMvcRequestBuilders.get("/users")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andReturn();
+        UserResponseDto[] userResponseDtos = objectMapper.readValue(result.getResponse().getContentAsString(), UserResponseDto[].class);
+
+        //then
+        assertEquals(savedUser.getId(), userResponseDtos[0].getId());
+        assertEquals(savedUser.getFirstName(), userResponseDtos[0].getFirstName());
+        assertEquals(savedUser.getLastName(), userResponseDtos[0].getLastName());
+    }
+
+    @Test
+    public void whenCreateUser_thenCorrect() throws Exception {
+
+        //given
+        BookDto book1 = new BookDto("Book1", "Author1");
+        BookDto book2 = new BookDto("Book2", "Author2");
+        UserCreationDto userCreationDto = new UserCreationDto("John", "Doe", "123 Main St", Arrays.asList(book1, book2));
+
+        //when
+        MvcResult result = mockMvc.perform(MockMvcRequestBuilders.post("/users")
+                        .content(asJsonString(userCreationDto))
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .accept(MediaType.APPLICATION_JSON))
+                .andReturn();
+        UserResponseDto userResponseDto = objectMapper.readValue(result.getResponse().getContentAsString(), UserResponseDto.class);
+
+        //then
+        assertEquals(userCreationDto.getFirstName(), userResponseDto.getFirstName());
+        assertEquals(userCreationDto.getLastName(), userResponseDto.getLastName());
+    }
+
+    // Helper method to convert objects to JSON
+    private String asJsonString(final Object obj) {
```

Comment:
Tests like this should not reuse the object mapper. We need hard coded assertions in order to actually cover e.g. the JSON field names.